### ISSUE TITLE
[boolector] boolector_sgte: sorts of 'e0' and 'e1' must match

### ISSUE THREAD

**Issue by lucasccordeiro** [esbmc/esbmc#148]
Command-line
$ esbmc queue.c 

Reduced C program:

```
extern int __VERIFIER_nondet_int(void);
extern void __VERIFIER_error() __attribute__ ((__noreturn__));
#include <pthread.h>
#define SIZE	(20)
#define EMPTY	(-1)
typedef struct {
    int element[SIZE];
    int head;
    int tail;
    int amount;
} QType;
pthread_mutex_t m;
int __VERIFIER_nondet_int();
int stored_elements[SIZE];
QType queue;
int empty(QType * q) {
  if (q->head == q->tail)  
    return EMPTY;
  else 
    return 0;
}
int enqueue(QType *q, int x) {
  q->element[q->tail] = x;
  q->amount++;
  if (q->tail == SIZE)
    q->tail = 1;
  else 
    q->tail++;
  return 0;
}
void *t1(void *arg) {
  int value, i;
  pthread_mutex_lock(&m);
  value = __VERIFIER_nondet_int();
  if (enqueue(&queue,value)) {
    goto ERROR;
  }
  stored_elements[0]=value;
  if (empty(&queue)) {
    goto ERROR;
  }
  pthread_mutex_unlock(&m);
  return NULL;
  ERROR: __VERIFIER_error();
}
int main(void) {
  pthread_t id1;
  pthread_create(&id1, NULL, t1, &queue);
  return 0;
}
```

Output:
```
Encoding remaining VCC(s) using bit-vector/floating-point arithmetic
Boolector error encountered
[boolector] boolector_sgte: sorts of 'e0' and 'e1' must match


Abort trap: 6
```

**Comment by lucasccordeiro** [comment]
I have just open a PR to fix this issue: https://github.com/esbmc/esbmc/pull/202

**Comment by mikhailramalho** [comment]
The following patch fixes the issue:
```
From 6d0e457ab19bc94c9f616ad6cf2bacbd3ea4e0df Mon Sep 17 00:00:00 2001
From: "Mikhail R. Gadelha" <mikhail.ramalho@gmail.com>
Date: Thu, 11 Jun 2020 18:46:03 -0400
Subject: [PATCH] Fix isse #148

Signed-off-by: Mikhail R. Gadelha <mikhail.ramalho@gmail.com>
---
 src/pointer-analysis/dereference.cpp | 10 +++++-----
 1 file changed, 5 insertions(+), 5 deletions(-)

diff --git a/src/pointer-analysis/dereference.cpp b/src/pointer-analysis/dereference.cpp
index 981b2f571..5b8476abb 100644
--- a/src/pointer-analysis/dereference.cpp
+++ b/src/pointer-analysis/dereference.cpp
@@ -1200,8 +1200,8 @@ void dereferencet::construct_from_dyn_struct_offset(
     // Round up to word size
     unsigned int word_mask = (config.ansi_c.word_size / 8) - 1;
     field_size = (field_size + word_mask) & (~word_mask);
-    expr2tc field_offs = gen_ulong(offs.to_uint64());
-    expr2tc field_top = gen_ulong(offs.to_uint64() + field_size);
+    expr2tc field_offs = constant_int2tc(offset->type, offs);
+    expr2tc field_top = constant_int2tc(offset->type, offs + field_size);
     expr2tc lower_bound = greaterthanequal2tc(offset, field_offs);
     expr2tc upper_bound = lessthan2tc(offset, field_top);
     expr2tc field_guard = and2tc(lower_bound, upper_bound);
@@ -2013,7 +2013,7 @@ void dereferencet::check_code_access(
 
     // Only other constraint is that the offset has to be zero; there are no
     // other rules about what code objects look like.
-    notequal2tc neq(offset, gen_ulong(0));
+    notequal2tc neq(offset, gen_zero(offset->type));
     guardt tmp_guard = guard;
     tmp_guard.add(neq);
     dereference_failure(
@@ -2072,12 +2072,12 @@ void dereferencet::check_alignment(
   {
     expr2tc align = gen_ulong(minwidth);
     modulus2tc moded(align->type, offset, align);
-    neq = notequal2tc(moded, gen_ulong(0));
+    neq = notequal2tc(moded, gen_zero(moded->type));
   }
   else
   {
     bitand2tc anded(mask_expr->type, mask_expr, offset);
-    neq = notequal2tc(anded, gen_ulong(0));
+    neq = notequal2tc(anded, gen_zero(anded->type));
   }
 
   guardt tmp_guard2 = guard;
-- 
2.25.1
```

However, there is a bigger issue here: the use of `gen_ulong` throughout ESBMC. I'll open another issue for it.

**Comment by lucasccordeiro** [comment]
Closed via https://github.com/esbmc/esbmc/pull/202.
